src/wallet/rpcwallet.cpp

```diff
@@ -2563,99 +2570,97 @@ Value zc_raw_pour(const json_spirit::Array& params, bool fHelp)
+    std::vector<JSInput> vpourin;
+    std::vector<JSOutput> vpourout;
+    std::vector<Note> notes;
+    std::vector<SpendingKey> keys;
+    std::vector<boost::optional<uint256>> commitments;
 
     BOOST_FOREACH(const Pair& s, inputs)
     {
-        CDataStream ssData(ParseHexV(s.name_, "bucket"), SER_NETWORK, PROTOCOL_VERSION);
-        uint64_t value;
-        std::vector<unsigned char> rho;
-        std::vector<unsigned char> r;
-
-        ssData >> value;
-        ssData >> rho;
-        ssData >> r;
-
-        uint256 a_sk;
-        uint256 sk_enc;
+        SpendingKey k;
 
         {
-            CDataStream ssData2(ParseHexV(s.value_, "zcsecretkey"), SER_NETWORK, PROTOCOL_VERSION);
+            CDataStream ssData(ParseHexV(s.value_, "zcsecretkey"), SER_NETWORK, PROTOCOL_VERSION);
             try {
-                ssData2 >> a_sk;
-                ssData2 >> sk_enc;
+                ssData >> k;
             } catch(const std::exception &) {
                 throw runtime_error(
                     "zcsecretkey could not be decoded"
                 );
             }
         }
 
-        libzerocash::PrivateAddress zcsecretkey(a_sk, sk_enc);
-        libzerocash::Address zcaddress(zcsecretkey);
-        libzerocash::Coin input_coin(zcaddress.getPublicAddress(), value, rho, r);
+        keys.push_back(k);
 
-        std::vector<unsigned char> commitment_v = input_coin.getCoinCommitment().getCommitmentValue();
-        uint256 commitment = uint256(commitment_v);
+        NotePlaintext npt;
 
-        libzcash::MerklePath path;
-        assert(pwalletMain != NULL);
-        if (!pwalletMain->WitnessBucketCommitment(commitment, path, anchor)) {
-            throw std::runtime_error("Couldn't find bucket in the blockchain");
+        {
+            CDataStream ssData(ParseHexV(s.name_, "bucket"), SER_NETWORK, PROTOCOL_VERSION);
+            ssData >> npt;
         }
 
-        vpourin.push_back(PourInput(input_coin, zcaddress, path));
+        PaymentAddress addr = k.address();
+        Note note = npt.note(addr);
+        notes.push_back(note);
+        commitments.push_back(note.cm());
     }
 
-    while (vpourin.size() < NUM_POUR_INPUTS) {
-        vpourin.push_back(PourInput(INCREMENTAL_MERKLE_TREE_DEPTH));
+    uint256 anchor;
+    std::vector<boost::optional<ZCIncrementalWitness>> witnesses;
+    pwalletMain->WitnessBucketCommitment(commitments, witnesses, anchor);
+
+    assert(witnesses.size() == notes.size());
+    assert(notes.size() == keys.size());
+
+    {
+        for (size_t i = 0; i < witnesses.size(); i++) {
+            if (!witnesses[i]) {
+                throw runtime_error(
+                    "pour input could not be found in tree"
+                );
+            }
+
+            vpourin.push_back(JSInput(*witnesses[i], notes[i], keys[i]));
+        }
+    }
+
+    while (vpourin.size() < ZC_NUM_JS_INPUTS) {
+        vpourin.push_back(JSInput());
     }
 
     BOOST_FOREACH(const Pair& s, outputs)
```

Comment:
(I don't fully understand why I wrote my original comment (it was during my pairing with Daira), I need to come back and re-review this code later.)